ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ExternalTableCopyTaskBuilder.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.hadoop.hive.shims.Utils;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Serializable;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+public class ExternalTableCopyTaskBuilder {
+  private static final Logger LOG = LoggerFactory.getLogger(ExternalTableCopyTaskBuilder.class);
+  private final ReplLoadWork work;
+  private final HiveConf conf;
+
+  ExternalTableCopyTaskBuilder(ReplLoadWork work, HiveConf conf) {
+    this.work = work;
+    this.conf = conf;
+  }
+
+  List<Task<? extends Serializable>> tasks(TaskTracker tracker) {
+    List<Task<? extends Serializable>> tasks = new ArrayList<>();
+    Iterator<DirCopyWork> itr = work.getPathsToCopyIterator();
+    while (tracker.canAddMoreTasks() && itr.hasNext()) {
+      DirCopyWork work = itr.next();
+      Task<DirCopyWork> task = TaskFactory.get(work, conf);
+      tasks.add(task);
+      tracker.addTask(task);
+      LOG.debug("added task for {}", work);
+    }
+    return tasks;
+  }
+
+  public static class DirCopyTask extends Task<DirCopyWork> implements Serializable {
+    private static final Logger LOG = LoggerFactory.getLogger(DirCopyTask.class);
+    private static final int MAX_COPY_RETRY = 5;
+
+    @Override
+    protected int execute(DriverContext driverContext) {
+      String distCpDoAsUser = conf.getVar(HiveConf.ConfVars.HIVE_DISTCP_DOAS_USER);
+
+      Path sourcePath = work.fullyQualifiedSourcePath;
+      Path targetPath = work.fullyQualifiedTargetPath;
+      if (conf.getBoolVar(HiveConf.ConfVars.REPL_ADD_RAW_RESERVED_NAMESPACE)) {
+        sourcePath = reservedRawPath(work.fullyQualifiedSourcePath.toUri());
+        targetPath = reservedRawPath(work.fullyQualifiedTargetPath.toUri());
+      }
+      int currentRetry = 0;
+      while (currentRetry < MAX_COPY_RETRY) {
+        try {
+          UserGroupInformation ugi = Utils.getUGI();
+          String currentUser = ugi.getShortUserName();
+          boolean usePrivilegedUser =
+              distCpDoAsUser != null && !currentUser.equals(distCpDoAsUser);
+
+          // do we create a new conf and only here provide this additional option so that we get away from
+          // differences of data in two location for the same directories ?
+          // basically add distcp.options.delete to hiveconf new object ?
+          FileUtils.distCp(
+              sourcePath.getFileSystem(conf), // source file system
+              Collections.singletonList(sourcePath),  // list of source paths
+              targetPath,
+              false,
+              usePrivilegedUser ? distCpDoAsUser : null,
+              conf,
+              ShimLoader.getHadoopShims());
+          return 0;
+        } catch (Exception e) {
+          if (++currentRetry < MAX_COPY_RETRY) {
+            LOG.warn("unable to copy", e);
```

Comment:
i dont think anyone is going to debug causes in between at those locations for WARN's since you will most probably need audit logs to see the history in time to identify what was the state of that location when it failed, if systems are build to recover with retries no point identifying that there was for ex a temporary network glitch